src/cunumeric/index/advanced_indexing_omp.cc

```diff
@@ -27,101 +27,90 @@ namespace cunumeric {
+                      const int volume,
+                      int64_t out_idx,
+                      const int key_dim,
+                      const int skip_size) const
   {
 #pragma omp for schedule(static)
     for (size_t idx = 0; idx < volume; ++idx) {
-      auto p = pitches_index.unflatten(idx, rect_index.lo);
+      auto p = pitches.unflatten(idx, rect.lo);
       if (index[p] == true) {
-        auto p_input = pitches_input.unflatten(idx, rect_input.lo);
-        out[out_idx] = input[p_input];
-        out_idx++;
+        Point<DIM> out_p;
+        out_p[0] = out_idx;
+        for (size_t i = 0; i < DIM - key_dim; i++) {
+          size_t j     = key_dim + i;
+          out_p[i + 1] = p[j];
+        }
+        for (size_t i = DIM - key_dim + 1; i < DIM; i++) out_p[i] = 0;
+        fill_out(out[out_p], p, in[p]);
+        if ((idx + 1) % skip_size == 0) out_idx++;
       }
     }
   }
 
-  void compute_output(Buffer<Point<DIM1>>& out,
-                      const AccessorRO<VAL, DIM1>&,
-                      const AccessorRO<bool, DIM2>& index,
-                      const Pitches<DIM1 - 1>& pitches_input,
-                      const Rect<DIM1>& rect_input,
-                      const Pitches<DIM2 - 1>& pitches_index,
-                      const Rect<DIM2>& rect_index,
-                      const size_t volume,
-                      int64_t out_idx) const
+  void operator()(Array& out_arr,
+                  const AccessorRO<VAL, DIM>& input,
+                  const AccessorRO<bool, DIM>& index,
+                  const Pitches<DIM - 1>& pitches,
+                  const Rect<DIM>& rect,
+                  const int key_dim) const
   {
-#pragma omp for schedule(static)
-    for (size_t idx = 0; idx < volume; ++idx) {
-      auto p = pitches_index.unflatten(idx, rect_index.lo);
-      if (index[p] == true) {
-        auto p_input = pitches_input.unflatten(idx, rect_input.lo);
-        out[out_idx] = p_input;
-        out_idx++;
-      }
+    size_t skip_size = 1;
+    for (int i = key_dim; i < DIM; i++) {
+      auto diff = 1 + rect.hi[i] - rect.lo[i];
+      if (diff != 0) skip_size *= diff;
     }
-  }
 
-  template <typename OUT_TYPE>
-  size_t operator()(Buffer<OUT_TYPE>& out,
-                    const AccessorRO<VAL, DIM1>& input,
-                    const AccessorRO<bool, DIM2>& index,
-                    const Pitches<DIM1 - 1>& pitches_input,
-                    const Rect<DIM1>& rect_input,
-                    const Pitches<DIM2 - 1>& pitches_index,
-                    const Rect<DIM2>& rect_index) const
-  {
-#ifdef DEBUG_CUNUMERIC
-    // in this case shapes for input and index arrays  should be the same
-    assert(Domain(rect_input) == Domain(rect_index));
-#endif
-    const size_t volume    = rect_index.volume();
     const auto max_threads = omp_get_max_threads();
+    const size_t volume    = rect.volume();
     size_t size            = 0;
     ThreadLocalStorage<int64_t> offsets(max_threads);
-
     {
```

Comment:
you may consider extracting this to a function (say `compute_output_offsets`) to document what it does.